### PR TITLE
Make VC++ version a parametrizable option for Windows CI.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32043 Make VC++ version a parametrizable option for Windows CI.**

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19621910](https://our.internmc.facebook.com/intern/diff/D19621910)

### CODE REVIEW

File: .jenkins/pytorch/win-test-helpers/build_pytorch.bat [hunk 1]
```diff
@@ -45,7 +45,7 @@ set CUDA_PATH_V9_2=%CUDA_PATH%
 goto cuda_build_common
 
 :cuda_build_10
-call "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.11
+call "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=%VC_VERSION%
```
Comment (peterjc123) [inline .jenkins/pytorch/win-test-helpers/build_pytorch.bat]: Actually, we can make 2017 and BuildTools as parameters too.
`VS_VERSION`: 2017
`VS_PRODUCT`: BuildTools

**Review by peterjc123 — APPROVED** [review]

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 0fb0c616:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=0fb0c616c8597bce77b69c210400c9a18d282ecd).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4201025) pytorch_windows_test2 (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83222958) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83222958))
<details>
<summary>
<code>C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(143): error C2872: 'std': ambiguous symbol</code>
</summary>

```
        [ 
            _Ty=int64_t 
        ] 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(141): error C2872: 'std': ambiguous symbol 
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\include\forward_list(20): note: could be 'std' 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(141): note: or       'std' 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(149): note: see reference to class template instantiation 'torch::autograd::CppNode<T>' being compiled 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(142): error C2872: 'std': ambiguous symbol 
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\include\forward_list(20): note: could be 'std' 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(142): note: or       'std' 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(143): error C2872: 'std': ambiguous symbol 
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\include\forward_list(20): note: could be 'std' 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch/csrc/autograd/custom_function.h(143): note: or       'std' 
C:\Users\circleci\project\build\win_tmp\build\torch\include\torch\csrc\api\include\torch/optim/optimizer.h(83): warning C4251: 'torch::optim::detail::OptimizerBase::parameters_': class 'std::vector<at::Tensor,std::allocator<_Ty>>' needs to have dll-interface to be used by clients of class 'torch::optim::detail::OptimizerBase' 
        with 
        [ 
            _Ty=at::Tensor 
        ] 
C:\Users\circleci\project\build\win_tmp\build\torch\include\ATen/core/TensorBody.h(479): note: see declaration of 'std::vector<at::Tensor,std::allocator<_Ty>>' 
        with 
        [ 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: .circleci/verbatim-sources/pytorch-job-specs.yml [hunk 2]
```diff
@@ -145,6 +145,7 @@ jobs:
       BUILD_ENVIRONMENT: "pytorch-win-ws2019-cuda10-cudnn7-py3"
       SCCACHE_BUCKET: "ossci-compiler-cache"
       CUDA_VERSION: "10"
+      VC_VERSION: "14.11"
```
Comment (peterjc123) [inline .circleci/verbatim-sources/pytorch-job-specs.yml]: Looks like you'll need to add them to the test job too.

**Review by peterjc123 — CHANGES_REQUESTED** [review]

**Comment by peterjc123** [comment]
@ezyang Do you plan to update this?

**Comment by ezyang** [comment]
Ya, doing so now.

**Comment by peterjc123** [comment]
Looks like the change is not working.

File: .circleci/verbatim-sources/pytorch-job-specs.yml [hunk 3]
```diff
@@ -177,16 +172,8 @@ jobs:
             set -x
             .jenkins/pytorch/win-build.sh
   pytorch_windows_test:
+    <<: *pytorch_windows_params
     executor: windows-with-nvidia-gpu
-    parameters:
```
Comment (peterjc123) [inline .circleci/verbatim-sources/pytorch-job-specs.yml]: Looks like we will need to keep some of the lines here.

**Review by peterjc123 — APPROVED** [review]

**Pull request closed without merging**